### PR TITLE
Assume config value to be json-formatted in cli

### PR DESCRIPTION
**Pull request by rht**
This deprecates explicit '--bool' and '--json' flag.
The reason is, if explicit type has to be specified, then '--integer',
'--float' also need to be implemented. But json parser already knows how
to handle these types, so why not just reuse it.
ref: #740

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 6c3357986bd0524f1e48e6bdb4d17021ac4949fa
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

Guidelines are available at: https://github.com/ipfs/community/blob/master/docs/commit-message.md Your feedback on GitCop is welcome on the following issue: https://github.com/ipfs/infrastructure/issues/23

---
This message was auto-generated by https://gitcop.com

**Comment by whyrusleeping** [comment]
we should re-run the tests here. everything appears to have failed

File: core/commands/config.go [hunk 1]
```diff
@@ -75,21 +75,14 @@ Set the value of the 'datastore.path' key:
...
-				output, err = setConfig(r, key, jsonVal)
-			} else if isbool, _, _ := req.Option("bool").Bool(); isbool {
-				output, err = setConfig(r, key, value == "true")
-			} else {
-				output, err = setConfig(r, key, value)
+			var jsonVal interface{}
+			if err := json.Unmarshal([]byte(value), &jsonVal); err != nil {
+				err = fmt.Errorf("failed to unmarshal json. %s", err)
+				res.SetError(err, cmds.ErrNormal)
+				return
```
Comment (jbenet) [inline core/commands/config.go]: so always default to json? that makes sense, but does that mean that we now have to do:

```
# used to work
> ipfs config Addresses.API /ip4/127.0.0.1/tcp/5001

# must now be
> ipfs config Addresses.API "/ip4/127.0.0.1/tcp/5001"
```
?

Comment (wking) [inline core/commands/config.go]: On Thu, Jun 11, 2015 at 01:34:07PM -0700, Juan Batiz-Benet wrote:
> # must now be
> > ipfs config Addresses.API "/ip4/127.0.0.1/tcp/5001"

Implementations differ in how strictly they enforce this, but I've
seen a number of folks suggest that strings are not valid JSON root
objects (e.g. [1]).  That's also how it's written up in RFC 4627 [2].

[1]: http://stackoverflow.com/a/10428530
[2]: https://tools.ietf.org/html/rfc4627#section-2

Comment (rht) [inline core/commands/config.go]: ```ipfs config Addresses.API /ip4/127.0.0.1/tcp/5001``` still works in this case (golang/pkg/encoding/json).

So I will have to check which implementation (RFC 4627 vs ECMA-262 5 vs ECMA-262 5.1) is among the majority today.

Comment (jbenet) [inline core/commands/config.go]: > `ipfs config Addresses.API /ip4/127.0.0.1/tcp/5001` still works

ok great.

**Comment by jbenet** [comment]
I re-ran the tests. something bad's going on.

- at least this: https://travis-ci.org/ipfs/go-ipfs/jobs/66392490#L400

File: core/commands/config.go (hunk 1 above)
Comment (rht) [inline core/commands/config.go]: accept stand-alone strings, numbers (RFC 7159):
- Go 1.3+
- Major browsers (IE 8+, everything else no IE)[1]
- Python 2.7+[1]

else (RFC 4627):
- Ruby 2.1.1 stdlib "json" [2]

Pending:
http://json.org/ (150-ish of those)

[1] http://stackoverflow.com/questions/18202532/can-a-single-stand-alone-literal-form-a-valid-json-document
[2] manual check & "JSON.generate only allows objects or arrays to be converted to JSON syntax." in http://ruby-doc.org/stdlib-2.0.0/libdoc/json/rdoc/JSON.html

---

Or since RFC 7159 supersedes RFC 4627, any legacy system will move toward it anyway.

**Comment by rht** [comment]
I see, I likely didn't type the beginning slash in testing ```ipfs config Addresses.API /ip4/127.0.0.1/tcp/5001```.

With this commit, all strings must be explicitly formatted as \"<string>\".

I suppose the options now are:
1. Default to string. Keep the flags as they are currently, then just add '--number'. Where '--bool' and '--number' are actually '--json' in disguise.
2. Default to string. Use '--json' for everything else (object, array, number, bool, null).

Option 2 is simpler so now I'm switching to it.

**Comment by jbenet** [comment]
- (1.) normally, i'd like this to be the end state, but this breaks user space. :(  not so sure.
- (2.) -1 dont like this one.
- (3.) I think i like this one. and deprecate `--bool`. (don't remove it yet, but later) 
- (4.) is a bit concerning, its the type of thing that backfires annoyingly. It may be ok we _also_ have `--json` and `--string` just in case the auto-detect fails.

**Comment by whyrusleeping** [comment]
maybe we can be a little smart about things. things like: if `strconv.Atoi` doesnt return an error, we can assume its a number

**Comment by jbenet** [comment]
I'd prefer being simple and explicit. Smartness can introduce weird edge cases



—

**Comment by rht** [comment]
I think ```strconv.Atoi``` is a valid literal type inference. Will this be extended to ```Atob```, ```Atof```, ```json.Unmarshal``` (for list and object)... then default to string? If so this is basically (4.).

- Default to string? (2, 3)
I have to amend one point in (3.): actually string is still the >50% most common value in the config (the reason why default to string is the current state in the first place). But other types are still common.

- Default to json? (1, 4)
There is no need to "maintain '--bool' backward compatibility", since bools/numbers are automatically parsed. One can still specify '--bool'/'--number' but this is placebo.

The only breaking change in (1.) is the necessity of "--string".
It's just that you rarely see this in other configuration languages. But you also rarely see explicit type annotation like '--bool', '--number'.

(4.) by itself blurs the edge cases. It emulates the flexibility "smartness" of a human parser, but also its sloppiness. But this can still be fine because the config struct's are explicitly typed. And so this will always detect a malformed bool/int/float/array/object as a final check. This 2nd type check doesn't exist for user-specified values, however (and if this is hard to debug...).

There is actually option (5.): use https://github.com/vstakhov/libucl (as a superset of json) just for parsing then serialize to json.
> There is no requirement of quotes for strings and keys, ...
It looks like strings can be inferred unambiguously and efficiently. But this is not a sufficient reason to integrate the whole library though (and also more formats to keep track of for compatibility: json, ucl, go types, other ipfs implementation types).

**Comment by jbenet** [comment]
@rht not sure here. i think we should make sure not to 
- break current user scripts (hence keep the `--bool` flag for now even if it doesnt do anything different)
- make sure not to get into weird edge cases.
- allow explicitness (make sure to keep `--json` which will _only_ parse json or fail)

if you want to try (4.) then let's do it, but we'll need good tests to make sure various things get set correctly to what we expect.

**Comment by rht** [comment]
The only edge case with (4) is with user-specified values.
Everything else is unchanged.

The `commands.Option` object actually has a very similar use case.
In this case, the choice is (4) + value auto-cast (https://github.com/ipfs/go-ipfs/blob/master/commands/request.go#L224-L267), which makes sense because there won't be any user-specified command options.
